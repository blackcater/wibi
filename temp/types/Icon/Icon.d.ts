import { VComponent } from '../component';

import { IconStack } from './IconStack';

export declare class Icon extends VComponent {
  static Stack: IconStack;

  /**
   * Icon name
   *
   * @default: ''
   **/
  icon: string | { viewBox: string; id: string };

  /**
   * Sets the icon size
   *
   * @default: ''
   **/
  size: 'lg' | '2x' | '3x' | '4x' | '5x';

  /**
   * Flip the icon
   *
   * @default: ''
   **/
  flip: 'horizontal' | 'vertical';

  /**
   * Combine multiple icons
   *
   * @default: ''
   **/
  stack: '1x' | '2x';

  /**
   * Rotate the icon
   *
   * @default: 0
   **/
  rotate: number;

  /**
   * Dynamic rotation icon
   *
   * @default: false
   **/
  spin: boolean;

  /**
   * Use pulse to have it rotate with 8 steps
   *
   * @default: false
   **/
  pulse: boolean;

  /**
   * Inverse color
   *
   * @default: false
   **/
  inverse: boolean;

  /**
   * Fixed icon width because there are many icons with uneven size
   *
   * @default: false
   **/
  fixedWidth: boolean;

  /**
   * Set SVG style when using custom SVG Icon
   *
   * @default: {}
   **/
  svgStyle: object;

  /**
   * You can use a custom element for this component
   *
   * @default: 'i'
   **/
  componentClass: string | object;
}
