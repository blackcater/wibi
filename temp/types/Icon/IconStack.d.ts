import { VComponent } from '../component';

export declare class IconStack extends VComponent {
  /**
   * Sets the icon size
   *
   * @default: ''
   **/
  size?: 'lg' | '2x' | '3x' | '4x' | '5x';

  /**
   * You can use a custom element for this component
   *
   * @default: 'span'
   **/
  componentClass: string | object;
}
