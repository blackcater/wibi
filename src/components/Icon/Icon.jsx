import VueTypes from 'vue-types';
import prefix, { defaultClassPrefix } from 'utils/prefix';

const CLASS_PREFIX = 'icon';

export default {
  name: 'Icon',

  props: {
    icon: VueTypes.oneOfType([VueTypes.string, VueTypes.object]),
    size: VueTypes.oneOf(['lg', '2x', '3x', '4x', '5x']),
    status: VueTypes.oneOf(['primary', 'success', 'info', 'warning', 'danger']),
    flip: VueTypes.oneOf(['horizontal', 'vertical']),
    stack: VueTypes.oneOf(['1x', '2x']),
    rotate: VueTypes.oneOf([0, 90, 180, 270]),
    spin: VueTypes.bool.def(false),
    pulse: VueTypes.bool.def(false),
    inverse: VueTypes.bool.def(false),
    fixedWidth: VueTypes.bool.def(false),
    svgStyle: VueTypes.object,
    classPrefix: VueTypes.string.def(defaultClassPrefix(CLASS_PREFIX)),
    componentClass: VueTypes.oneOfType([VueTypes.string, VueTypes.object]).def(
      'i'
    ),
  },

  computed: {
    isSvgIcon() {
      return typeof this.icon === 'object' && this.icon.id && this.icon.viewBox;
    },

    classes() {
      return [
        this.classPrefix,
        {
          [this._addPrefix(this.icon)]: !this.isSvgIcon,
          [this._addPrefix(this.size)]: this.size,
          [prefix('text', this.status)]: this.status,
          [this._addPrefix('spin')]: this.spin,
          [this._addPrefix('pulse')]: this.pulse,
          [this._addPrefix('inverse')]: this.inverse,
          [this._addPrefix('fw')]: this.fixedWidth,
          [this._addPrefix(`flip-${this.flip || ''}`)]: this.flip,
          [this._addPrefix(`stack-${this.stack || ''}`)]: this.stack,
          [this._addPrefix(`rotate-${this.rotate || ''}`)]: this.rotate,
        },
      ];
    },
  },

  render() {
    const Component = this.componentClass;
    const iconData = {
      class: this.classes,
      props: this.attrs,
      on: this.$listeners,
    };

    return (
      <Component {...iconData}>
        {this.isSvgIcon ? (
          <svg style={this.svgStyle} viewBox={this.icon.viewBox}>
            <use xlinkHref={`#${this.icon.id}`} />
          </svg>
        ) : null}
      </Component>
    );
  },

  methods: {
    _addPrefix(cls) {
      return prefix(this.classPrefix, cls);
    },
  },
};
