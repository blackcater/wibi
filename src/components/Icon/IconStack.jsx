import VueTypes from 'vue-types';
import prefix, { defaultClassPrefix } from 'utils/prefix';

const CLASS_PREFIX = 'icon';

export default {
  name: 'IconStack',

  props: {
    size: VueTypes.oneOf(['lg', '2x', '3x', '4x', '5x']),
    status: VueTypes.oneOf(['primary', 'success', 'info', 'warning', 'danger']),
    flip: VueTypes.oneOf(['horizontal', 'vertical']),
    rotate: VueTypes.oneOf([0, 90, 180, 270]),
    spin: VueTypes.bool.def(false),
    pulse: VueTypes.bool.def(false),
    inverse: VueTypes.bool.def(false),
    fixedWidth: VueTypes.bool.def(false),
    classPrefix: VueTypes.string.def(defaultClassPrefix(CLASS_PREFIX)),
    componentClass: VueTypes.oneOfType([VueTypes.string, VueTypes.object]).def(
      'span'
    ),
  },

  computed: {
    classes() {
      const addPrefix = prefix(this.classPrefix);

      return [
        prefix(this.classPrefix, 'stack'),
        {
          [addPrefix(this.size)]: this.size,
          [prefix('text', this.status)]: this.satus,
          [addPrefix('spin')]: this.spin,
          [addPrefix('pulse')]: this.pulse,
          [addPrefix('inverse')]: this.inverse,
          [addPrefix('fw')]: this.fixedWidth,
          [addPrefix(`flip-${this.flip || ''}`)]: this.flip,
          [addPrefix(`rotate-${this.rotate || ''}`)]: this.rotate,
        },
      ];
    },
  },

  render() {
    const Component = this.componentClass;
    const iconStackData = {
      class: this.classes,
      props: this.$attrs,
      on: this.$listeners,
    };

    return <Component {...iconStackData}>{this.$slots.default}</Component>;
  },
};
