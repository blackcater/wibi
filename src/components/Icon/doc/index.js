export default [
  'basic',
  'spin',
  'rotate',
  'size',
  'stack',
  'custom',
  'custom-svg',
];
