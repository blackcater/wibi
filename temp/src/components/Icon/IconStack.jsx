import VueTypes from 'vue-types';
import prefix, { defaultClassPrefix } from 'utils/prefix';
import { splitDataByComponent } from 'utils/split';

const CLASS_PREFIX = 'icon-stack';

export default {
  name: 'IconStack',

  props: {
    size: VueTypes.oneOf(['lg', '2x', '3x', '4x', '5x']),

    classPrefix: VueTypes.string.def(defaultClassPrefix(CLASS_PREFIX)),
    componentClass: VueTypes.oneOfType([VueTypes.string, VueTypes.object]).def(
      'span'
    ),

    // slot
  },

  computed: {
    classes() {
      return [
        this.classPrefix,
        {
          [this._addPrefix(`size-${this.size || ''}`)]: this.size,
        },
      ];
    },
  },

  render() {
    const Component = this.componentClass;
    const iconStackData = splitDataByComponent(
      {
        class: this.classes,
        splitProps: this.$attrs,
        on: this.$listeners,
      },
      Component
    );

    return <Component {...iconStackData}>{this.$slots.default}</Component>;
  },

  methods: {
    _addPrefix(cls) {
      return prefix(this.classPrefix, cls);
    },
  },
};
