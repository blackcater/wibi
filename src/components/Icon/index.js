import Icon from './Icon.jsx';
import IconStack from './IconStack.jsx';

export { IconStack };
export default Icon;
