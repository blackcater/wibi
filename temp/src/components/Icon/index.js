import Icon from './Icon.jsx';
import IconStack from './IconStack.jsx';

Icon.Stack = IconStack;

Icon.install = function(Vue) {
  Vue.component(Icon.name, Icon);
  Vue.component(IconStack.name, IconStack);
};

export default Icon;
