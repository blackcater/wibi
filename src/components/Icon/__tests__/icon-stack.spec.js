import { mount } from '@vue/test-utils';
import Icon from 'components/Icon';

const IconStack = Icon.Stack;

describe('Button component', () => {
  it('renders correctly', () => {
    const wrapper = mount({
      render() {
        return (
          <IconStack size="lg">
            <Icon icon="square" stack="2x" />
            <Icon icon="terminal" stack="1x" inverse />
          </IconStack>
        );
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
