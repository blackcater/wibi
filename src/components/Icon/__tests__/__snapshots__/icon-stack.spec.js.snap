// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders correctly 1`] = `<span class="vs-icon-stack vs-icon-stack-size-lg"><i class="vs-icon vs-icon-square vs-icon-stack-2x"></i><i class="vs-icon vs-icon-terminal vs-icon-inverse vs-icon-stack-1x"></i></span>`;
